Add note in breaking changes for nameid_format (#77785)
elastic/elasticsearch: docs/reference/migration/migrate_8_0/security.asciidoc

result in an error on startup.
====

[discrete]
[[saml-realm-nameid-changes]]

> .The `nameid_format` SAML realm setting has no default value.

Reviewer: Should this be:
```suggestion
.The `nameid_format` SAML realm setting no longer has a default value.
```
?

Our breaking changes are usually described as a _change_ rather than a statement of the new behaviour.
That tends to be easier for people who are upgrading because it makes it easier to understand what is different and how that might affect them.

Author: SGTM

[%collapsible]
====
*Details* +
In SAML, Identity Providers (IdPs) can either be explicitly configured to
release a `NameID` with a specific format, or configured to attempt to conform 
with the requirements of a Service Provider (SP). The SP declares its
requirements in the `NameIDPolicy` element of a SAML Authentication Request.
In {es}, the `nameid_format` SAML realm setting controls the `NameIDPolicy`
value.

Previously, the default value for `nameid_format` was
`urn:oasis:names:tc:SAML:2.0:nameid-format:transient`. This setting created
authentication requests that required the IdP to release `NameID` with a
`transient` format.

The default value has been removed, which means that {es} will create SAML Authentication Requests by default that don't put this requirement on the
IdP. If you want to retain the previous behavior, set `nameid_format` to
`urn:oasis:names:tc:SAML:2.0:nameid-format:transient`.

*Impact* +
If you currently don't configure `nameid_format` explicitly, it's possible

> that your configuration won't match the `NameID` format that your IdP expects.

Reviewer 2: Sorry for drive-by comment but should this be
```suggestion
that your configuration won't match the `NameID` format that your IdP releases.
```
?
My understanding is that if `nameid_format` is not explicitly configured, in a newer version of ES, the idP is free to release a `NameID` with a format by its own choice and this format may not work with the ES SAML configuration?

Reviewer 3: @ywangd, I think you're correct in using "releases" here. Looking at the [SAML troubleshooting guide](https://www.elastic.co/guide/en/elasticsearch/reference/7.14/trb-security-saml.html), we indicate that:

> This is controlled by the nameid_format configuration parameter in elasticsearch.yml, which if not set defaults to urn:oasis:names:tc:SAML:2.0:nameid-format:transient. This instructs the Identity Provider to return a NameID with that specific format in the SAML Response. If the SAML Identity Provider cannot grant that request, for example because it is configured to release a NameID format with urn:oasis:names:tc:SAML:2.0:nameid-format:persistent format instead, it returns this error indicating an invalid NameID policy.

Author: > My understanding is that if nameid_format is not explicitly configured, in a newer version of ES, the idP is free to release a NameID with a format by its own choice and this format may not work with the ES SAML configuration?

`nameid_format` doesn't configure what should be matched in what the IDP releases though, so why does `release` make more sense here?

In any case ,we can only exactly match NameID with persistent format, we have the `nameid:persistent` magic string value that matches the NameID in the SAML response, but only if it is with persistent format. As such if you were already using 
```
atttributes.principal: nameid:persistent
```

then you either have already set `nameid_format: "urn:oasis:names:tc:SAML:2.0:nameid-format:persistent"` in your config, or your IDP is configured to release a certain format no matter what. In both these cases , your setup will continue to work. 

I chose `expects` on purpose here to capture the possibility that the IDP is configured to release a NameID format based on what's in the NameID Policy (and it's own constraints), and the default change causes the SP to request a different format so the IDP cannot grant that request (because it violates it's own constraints) 

@lockewritesdocs I don't think there is something in this text that sways us one way or the other here, but it's a good reminder that this text is now stale and we need to update it for 8.0.0

Reviewer 2: Thanks @jkakavas for the explanation. Apperantly my previous understanding on how idP responds to NameID request is wrong. Just one follow-up question on:

> I chose expects on purpose here to capture the possibility that the IDP is configured to release a NameID format based on what's in the NameID Policy (and it's own constraints), and the default change causes the SP to request a different format so the IDP cannot grant that request (because it violates it's own constraints)

ES's previous default for nameid is `transient`, now it is removed, which means a request from ES will not require certain nameid format. Are you saying that some idP could fail to respond when the SP does not explicitly specify a nameid format?

Reviewer: I find this wording to be strange as well. 

> If you currently don't configure `nameid_format` explicitly, it's possible
> that your configuration won't match the `NameID` format that your IdP expects.

_"If you currently don't configure `nameid_format` explicitly"_ then the behaviour will change from setting a NameIDPolicy with a value of transient, to no longer setting a NameIDPolicy.

If the IdP is expecting a NameIDPolicy, then this will be a problem. 
But this _"your configuration won't match the `NameID` format that your IdP expects"_ doesn't read like that because it implies that the configuration has a "`NameID` format", when the problem (if there is one) will be that the configuration lacks a "`NameID` format".

If I understand your intent correctly, I think we want something like:
```suggestion
that your IdP will reject authentication request from {es} because the requests 
do not specify a `NameID` format (and your IdP is configured to expect one).
```

Author: I should rewrite 

> I chose expects on purpose here to capture the possibility that the IDP is configured to release a NameID format based on what's in the NameID Policy (and it's own constraints), and the default change causes the SP to request a different format so the IDP cannot grant that request (because it violates it's own constraints)

to 

> I chose expects on purpose here to capture the possibility that the IDP is configured to release a NameID format based on what's in the NameID Policy (and it's own constraints), and the default change causes the SP to not request a specific format so the IDP cannot grant that request because it has no fallback/default for what format to use.

I'll flick you a link from a slack thread where I had a contest with Tim about "what's the most contrived example this could fail in".

As said, this is a contrived example, I know of no IDP implementations that behave this way. Most IDPs 

1) Either require the admin to configure the format of the NameID it should use and do not care about what's in the policy of the request ( vast majority )
2) or allow a set of formats to be configured and limit that by what is in the policy of the request

Author: Yes Tim that's the intent and I like your wording, I will adopt.

> This mismatch can result in a broken SAML configuration. If you're unsure which 
> name format your IdP uses to generate a value for `NameID`, try setting
> `nameid_format` to `urn:oasis:names:tc:SAML:2.0:nameid-format:transient` 
> explicitly.

Author: ```suggestion
This mismatch can result in a broken SAML configuration. If you're unsure whether 
your IdP is explicitly configured to use a certain `NameID` format and you want to retain current behavior
, try setting `nameid_format` to `urn:oasis:names:tc:SAML:2.0:nameid-format:transient` explicitly.
```

result in an error on startup.
====

[discrete]
[[saml-realm-nameid-changes]]
.The `nameid_format` SAML realm setting no longer has a default value.
[%collapsible]
====
*Details* +
In SAML, Identity Providers (IdPs) can either be explicitly configured to
release a `NameID` with a specific format, or configured to attempt to conform 
with the requirements of a Service Provider (SP). The SP declares its
requirements in the `NameIDPolicy` element of a SAML Authentication Request.
In {es}, the `nameid_format` SAML realm setting controls the `NameIDPolicy`
value.

Previously, the default value for `nameid_format` was
`urn:oasis:names:tc:SAML:2.0:nameid-format:transient`. This setting created
authentication requests that required the IdP to release `NameID` with a
`transient` format.

The default value has been removed, which means that {es} will create SAML Authentication Requests by default that don't put this requirement on the
IdP. If you want to retain the previous behavior, set `nameid_format` to
`urn:oasis:names:tc:SAML:2.0:nameid-format:transient`.

*Impact* +
If you currently don't configure `nameid_format` explicitly, it's possible

> that your IdP will reject authentication request from {es} because the requests 

Reviewer 3: ```suggestion
that your IdP will reject authentication requests from {es} because the requests 
```

*Impact* +
Assign users with the `kibana_user` role to the `kibana_admin` role.
Discontinue use of the `kibana_user` role.
====

[discrete]
[[saml-realm-nameid-changes]]
===== The default value of `nameid_format` setting has been removed.

.The `nameid_format` SAML realm setting has no default value.
[%collapsible]
====
*Details* +

> In SAML, Identity Providers (IdPs) either release a `NameID` or attempt to
> conform with the requirements of a Service Provider (SP). The SP declares its
> requirements in the `NameIDPolicy` of an authentication request. In {es}, the
> `nameid_format` SAML realm setting controls the `NameIDPolicy`.

Author: ```suggestion
In SAML,  Identity Providers (IdPs) can be either statically configured to release a `NameID`
with a specific format, or configured to try to conform with the requirements of Service Provider (SP)
The SP declares its requirements in the `NameIDPolicy` element of a SAML Authentication Request.
In {es}, the `nameid_format` SAML realm setting controls the `NameIDPolicy` value.
```

Previously, the default value for `nameid_format` was
`urn:oasis:names:tc:SAML:2.0:nameid-format:transient`. This default created
authentication requests that would require the IdP to release `NameID` with a
transient format.

> The default value has now been removed. This means that {es} will be default

Author: ```suggestion
The default value has now been removed. This means that {es} will by default
```

create SAML Authentication Requests that do not put forward such requirements
to the Identity Provider.

If you want to retain the previous behavior, you can set `nameid_format`
to `urn:oasis:names:tc:SAML:2.0:nameid-format:transient`.

*Impact* +

> To avoid issues, explicitly configure `nameid_format`. If you don't configure

Author: This is not what I am trying to say. If we instruct everyone to set this value, we might as well leave the default in place. What I am trying to say is that "this should probably be ok, but if you know you want the old behavior or you don't know how your IDP is configured but it was working so far and you don't care to try, set this setting to this value explicitly"

*Impact* +
Assign users with the `kibana_user` role to the `kibana_admin` role.
Discontinue use of the `kibana_user` role.
====

[discrete]
[[saml-realm-nameid-changes]]

> ===== The default value of `nameid_format` setting has been removed.
> 

Reviewer 4: I'd remove this heading (keep the anchor) and relocate these changes just below the section for `The transport.profiles.*.xpack.security.type setting has been removed.`
```suggestion
```

Reviewer 3: Agreed! I'll implement that change.

> .The `nameid_format` setting of SAML realms has no default value.

Reviewer 4: Minor readability nit.
```suggestion
.The `nameid_format` SAML realm setting has no default value.
```

[%collapsible]
====
*Details* +

> Identity Providers can be either statically configured to release a `NameID`
> with a specific format, or they can be configured to try to conform with the
> requirements of the SP. The SP declares its requirements as part of the
> Authentication Request, using an element which is called the `NameIDPolicy`
> For the {es} SAML realm, `NameIDPolicy` is controlled by `nameid_format` in
> SAML realm settings.

Reviewer 4: I tried to streamline this and remove some passive voice. I'd also ensure we use acronyms like IdPs and SPs consistently.

I'm not as familiar with the audience here, but I wondered if we needed any of the context before the last sentence. Would the audience already be familiar with these concepts?
```suggestion
In SAML, Identity Providers (IdPs) either release a `NameID` or attempt to
conform with the requirements of a Service Provider (SP). The SP declares its
requirements in the `NameIDPolicy` of an authentication request. In {es}, the
`nameid_format` SAML realm setting controls the `NameIDPolicy`.
```

> Previously, the default value for this setting was

Reviewer 4: ```suggestion
Previously, the default value for `nameid_format` was
```

> `urn:oasis:names:tc:SAML:2.0:nameid-format:transient` which meant that {es}
> would create SAML Authentication Requests that would require from the Identity
> Provider to release `NameID` with a transient format.

Reviewer 4: I'd break this sentence up and use the `IdP` acronym.
```suggestion
`urn:oasis:names:tc:SAML:2.0:nameid-format:transient`. This default created
authentication requests that would require the IdP to release `NameID` with a
transient format.
```

> The default value has now been removed. This means that {es} will be default
> create SAML Authentication Requests that do not put forward such requirements
> to the Identity Provider.
> 
> If you want to retain the previous behavior, you can set `nameid_format`
> to `urn:oasis:names:tc:SAML:2.0:nameid-format:transient`.

Reviewer 4: I wonder if we can remove the second sentence.
```suggestion
This default has been removed. To retain the previous default behavior, set
`nameid_format` to `urn:oasis:names:tc:SAML:2.0:nameid-format:transient`.
```

Author: I don't think it makes sense to remove this. This is the essence of what changes and what this change means.

Reviewer 3: +1 to this change -- I'll include it in the next push.

*Impact* +

> If you are not configuring `nameid_format` explicitly in
> your current configuration, there might be cases where the combination of this
> configuration and the configuration of your SAML Identity Provider will result
> in a non-working setup. If you are uncertain of how your Identity Provider is
> configured to generate a value for `NameID`, you can set `nameid_format`
> to `urn:oasis:names:tc:SAML:2.0:nameid-format:transient` explicitly.

Reviewer 4: Tried to streamline this a bit.
```suggestion
To avoid issues, explicitly configure `nameid_format`. If you don't configure
`nameid_format`, some SAML setups may no longer work. If you don't know
how your IdP generates a value for `NameID`, explicitly set `nameid_format` to
`urn:oasis:names:tc:SAML:2.0:nameid-format:transient`.
```
